### PR TITLE
sort now prioritizes exact matches

### LINKED ISSUE #3: Rank exact matches above partial matches
Example: https://chinesedict2.web.app/?q=bus

This is happening because all search results are ranked by word frequency, which promotes partial matches with higher frequency above exact matches with lower frequency.

### PR DESCRIPTION
**Pull request by jagLas**
Fixes #3
The sorting algorithm first prioritizes if the searchable string is an exact match, then for matches that contain the string, then by frequency.

### CODE REVIEW

**Comment by jeffcarp** [comment]
Thanks! When I get a sec let me verify locally before merging, looks good!

Now that the sort logic is getting more complicated, as a followup it would be great to set up some unit tests for them. Lmk if you're interested.

File: public/main.js
Comment (jeffcarp) [inline public/main.js:72]: Could you also add exact matching support for definitions? That one might be trickier because the definitions are newline-separated. So you could match on each item from `a.definition.split('\n')` - though that might be slow enough to warrant a more elegant search method.

**Review by jeffcarp — CHANGES_REQUESTED** [review]

File: public/main.js
Comment (jagLas) [reply public/main.js:72]: So I added simple logic to see if the definitions include the string phrase by splitting the definitions, and it is a little slow for inputs of 2 or fewer characters. Inputs of 3 or higher seem to have the results filtered enough that it can sort them quickly enough to feel responsive.

I'll think about alternatives or other solutions. Maybe adding some kind of cache would help so that it doesn't have to re parse the dictionaries every time it's sorting.

**Comment by jagLas** [comment]
I made the requested changes and used a cache to speed things up a little bit. Chrome seems responsive for all length entries. Firefox has a tiny hang with 1 letter, but has sped up on 2 letter or more.

File: public/main.js
Comment (jagLas) [reply public/main.js:96]: I made the requested changes and used a cache to speed things up a little bit. Chrome seems responsive for all length entries. Firefox has a tiny hang with 1 letter, but has sped up on 2 letter or more.

**Comment by jagLas** [comment]
Let me know if this is good to go now or if there is anything else you would like addressed.

**Comment by jagLas** [comment]
I did some rudimentary bench marking and I'm getting some strange results and wanted your input. Adding a cache for the split dictionaries makes the search almost 2.5 times faster in chrome. For some reason, however, it makes it slower by about 30% in firefox, and I'm not sure why that is the case. So, for example, searching the letter p takes 173ms in firefox without a cache, and brings it up to 225ms with one. In chrome, however, searching the same letter takes about 90ms without a cache and 34 with one (after the first search).

File: public/main.js
Comment (jeffcarp) [inline public/main.js:4]: Thanks, I think this is the right direction. Can you leave a TODO here to replace this with a precomputed and more search-efficient data structure (e.g. a trie)?

**Review by jeffcarp — APPROVED** [review]
Thanks a ton! Looks good to go. I think we're at the point where it makes sense to rethink the data structure used for search. I'll file a bug to follow up on that.

> I did some rudimentary bench marking and I'm getting some strange results and wanted your input. Adding a cache for the split dictionaries makes the search almost 2.5 times faster in chrome. For some reason, however, it makes it slower by about 30% in firefox, and I'm not sure why that is the case. So, for example, searching the letter p takes 173ms in firefox without a cache, and brings it up to 225ms with one. In chrome, however, searching the same letter takes about 90ms without a cache and 34 with one (after the first search).

That is super weird. Since this is still an improvement in functionality let's move forward with this and we can solve the latency issue by replacing the dictionary array with a more appropriate data structure.

**Comment by jeffcarp** [comment]
Going to merge so I can deploy, we can work on it from there. Thanks for this!

**Pull request merged**
